node/impls.go

```diff
@@ -161,108 +162,29 @@ func (l agreementLedger) EnsureValidatedBlock(ve agreement.ValidatedBlock, c agr
-	logging.Base().Debug("consensus was reached on a block we don't have yet: ", blockHash)
-	for {
-		// Ask the fetcher to get the block somehow
-		block, fetchedCert, err := l.FetchBlockByDigest(round, quit)
-		if err != nil {
-			select {
-			case <-quit:
-				logging.Base().Debugf("EnsureDigest was asked to quit before we could acquire the block")
-				return
-			default:
-			}
-			logging.Base().Panicf("EnsureDigest could not acquire block, fetcher errored out: %v", err)
-		}
-
-		if block.Hash() == blockHash && block.ContentsMatchHeader() {
-			l.EnsureBlock(block, cert)
-			return
-		}
-		// Otherwise, fetcher gave us the wrong block
-		logging.Base().Warnf("fetcher gave us bad/wrong block (for round %d): fetched hash %v; want hash %v", round, block.Hash(), blockHash)
-
-		// As a failsafe, if the cert we fetched is valid but for the wrong block, panic as loudly as possible
-		if cert.Round == fetchedCert.Round &&
-			cert.Proposal.BlockDigest != fetchedCert.Proposal.BlockDigest &&
-			fetchedCert.Authenticate(block, l, verifier) == nil {
-			s := "!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!\n"
-			s += "!!!!!!!!!! FORK DETECTED !!!!!!!!!!!\n"
-			s += "!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!\n"
-			s += "EnsureDigest called with a cert authenticating block with hash %v.\n"
-			s += "We fetched a valid cert authenticating a different block, %v. This indicates a fork.\n\n"
-			s += "Cert from our agreement service:\n%#v\n\n"
-			s += "Cert from the fetcher:\n%#v\n\n"
-			s += "Block from the fetcher:\n%#v\n\n"
-			s += "!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!\n"
-			s += "!!!!!!!!!! FORK DETECTED !!!!!!!!!!!\n"
-			s += "!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!\n"
-			s = fmt.Sprintf(s, cert.Proposal.BlockDigest, fetchedCert.Proposal.BlockDigest, cert, fetchedCert, block)
-			fmt.Println(s)
-			logging.Base().Error(s)
-		}
+	waitForRound := l.Wait(cert.Round)
+	// clear out the pending certificates ( if any )
+	select {
+	case pendingCert := <-l.UnmatchedPendingCertificates:
+		logging.Base().Debugf("agreementLedger.EnsureDigest has flushed out pending request for certificate for round %d in favor of recent certificate for round %d", pendingCert.Cert.Round, cert.Round)
+	default:
 	}
-}
 
-func (l agreementLedger) innerFetch(fetcher rpcs.Fetcher, round basics.Round, quit chan struct{}) (*bookkeeping.Block, *agreement.Certificate, error) {
-	ctx, cancel := context.WithTimeout(context.Background(), rpcs.DefaultFetchTimeout)
-	defer cancel()
-	type fbreturn struct {
-		block *bookkeeping.Block
-		cert  *agreement.Certificate
-		err   error
-	}
-	localdone := make(chan fbreturn, 1)
-	go func() {
-		block, cert, _, err := fetcher.FetchBlock(ctx, round)
-		localdone <- fbreturn{block, cert, err}
-	}()
 	select {
-	case ret := <-localdone:
-		return ret.block, ret.cert, ret.err
+	case l.UnmatchedPendingCertificates <- catchup.PendingUnmatchedCertificate{Cert: cert, VoteVerifier: verifier}:
+		// good, we've placed the cert in the queue.
 	case <-quit:
```

Comment:
Given that this channel send will be non-blocking, do we need to handle this?

Reply:
Also nit: you should comment that the channel send is guaranteed to be non-blocking.

Reply:
sure.

Reply:
Given that the channel send is non-blocking, it might be worth ignoring the `quit` channel here since we'll `select` on it immediately afterwards anyway.

Reply:
The intent here was to avoid writing the content to the channel is quit is closed. I believe it's still correct, although the implementation is incorrect. I'll fix it.

Reply:
I'm confused--this doesn't seem to solve the problem? Since `s.quit` can always be closed after this `select` block, which races on the channel send below...

Reply:
The `quit` can get closed at any moment. that's true.
When the `quit` is closed after/during the send into the channel then the logic on the catchup would need to tackle with that ( I believe that it does ). However, in the case the `EnsureDigest` was called with a closed quit channel, there is no reason to push that onto the channel. We want to expedite the shutdown process.

Reply:
I guess the check seems unnecessary to me, and it's unclear how much it matters in practice. This isn't a blocker for the PR though.